Content/NPCs/Dungeon/CrescentCaster.cs

```diff
@@ -0,0 +1,505 @@
+        public float fade;
+
+        public float resetCounter = 0;
+        public float resetCounterIncrement;
+
+        public bool cloneCreated = false;
+
+        public Color baseColor = new Color(200, 230, 255);
+        public Color endColor = Color.Purple;
+
+        public float distanceFade => 1 - (resetCounter / 30f);
+
+        public CrescentCasterBolt(NPC targetNPC, NPC owner, Vector2 midPoint, Vector2 midPointDirection, GraphicsDevice device)
+        {
+            TargetNPC = targetNPC;
+            Owner = owner;
+            MidPoint = midPoint;
+            MidPointDirection = midPointDirection;
+            resetCounterIncrement = Main.rand.NextFloat(0.85f, 1.15f);
+
+            trail = new Trail(device, 15, new TriangularTip(4), factor => 16, factor =>
+            {
+                if (factor.X > 0.99f)
+                    return Color.Transparent;
+
+                return new Color(160, 220, 255) * fade * 0.1f * EaseFunction.EaseCubicOut.Ease(1 - factor.X) * distanceFade;
+            });
+
+            trail2 = new Trail(device, 15, new TriangularTip(4), factor => 3 * Main.rand.NextFloat(0.55f, 1.45f), factor =>
+            {
+                float progress = EaseFunction.EaseCubicOut.Ease(1 - factor.X);
+                return Color.Lerp(baseColor, endColor, EaseFunction.EaseCubicIn.Ease(1 - progress)) * fade * progress * distanceFade;
+            });
+
+            UpdateTrailPoints();
+
+            fade = 0;
+        }
+
+        public void UpdateTrailPoints()
+        {
+            List<Vector2> points = BezierPoints();
+
+            List<Vector2> pointsWithOffset = new List<Vector2>();
+
+            int index = 0;
+            foreach(Vector2 point in points)
+            {
+                float offsetAmount;
+                if (index == 0)
+                    offsetAmount = 0;
+                else
+                    offsetAmount = (point - points[index - 1]).Length() * 0.5f * (1.125f - (Math.Abs(7 - index) / 7f));
+
+                    index++;
+                pointsWithOffset.Add(point + Main.rand.NextVector2Circular(offsetAmount, offsetAmount));
+            }
+
+            trail.Positions = trail2.Positions = pointsWithOffset.ToArray();
+            trail.NextPosition = trail2.NextPosition = TargetNPC.Center;
+        }
+
+        public List<Vector2> BezierPoints()
+        {
+            Vector2 startPoint = Owner.Center + new Vector2(Owner.spriteDirection * 3, -25);
+            BezierCurve curve = new BezierCurve(startPoint, MidPoint, TargetNPC.Center);
+            return curve.GetPoints(15);
+        }
+
+        public void Render(Effect effect)
+        {
+            trail?.Render(effect);
+            trail2?.Render(effect);
+        }
+    }
+
+    internal class CrescentCaster : ModNPC,IDrawPrimitive
+    {
+        public override string Texture => AssetDirectory.DungeonNPC + Name;
+
+        private const int XFRAMES = 3;
+        private readonly float ACCELERATION = 0.15f;
```

Comment:
should these be constant? The capitilzation scheme seems to imply it but they lack the const keyword

Reply:
fixed